samples/s3/src/main/java/com/example/s3/PutGet.java

```diff
@@ -0,0 +1,147 @@
+                + "    throughput: Target throughput in gigabits per second (default 25.0)\n\n"
+                + "    part-size: Size in bytes of each multipart upload (default 8MB)\n\n";
+        if (args.length < 4) {
+            System.out.print(USAGE);
+            System.exit(1);
+        }
+
+        String bucket = args[0];
+        String key = args[1];
+        String putFilepath = args[2];
+        String getFilepath = putFilepath + ".download";
+        String signingRegion = args[3];
+
+        long partSizeBytes = 8 * 1024 * 1024; // 8MB
+        if (args.length > 4)
+            partSizeBytes = Long.parseLong(args[4]);
+
+        double targetThroughputGbps = 25.0;
+        if (args.length > 5)
+            targetThroughputGbps = Double.parseDouble(args[5]);
+
+        Log.initLoggingToStderr(Log.LogLevel.Error);
+
+        // try-with-resources to ensure these resources get closed
+        try (EventLoopGroup eventLoopGroup = new EventLoopGroup(0 /* cpuGroup */, 0 /* numThreads 0=default */);
+                HostResolver hostResolver = new HostResolver(eventLoopGroup, 256 /* maxEntries */);
+                ClientBootstrap clientBootstrap = new ClientBootstrap(eventLoopGroup, hostResolver);
+
+                // default provider searches for AWS credentials in places like
+                // ~/.aws/credentials and env vars AWS_ACCESS_KEY_ID/AWS_SECRET_ACCESS_KEY
+                CredentialsProvider credentialsProvider = new DefaultChainCredentialsProvider.DefaultChainCredentialsProviderBuilder()
+                        .withClientBootstrap(clientBootstrap).build();
+
+                S3NativeClient s3Client = new S3NativeClient(signingRegion, clientBootstrap, credentialsProvider,
+                        partSizeBytes, targetThroughputGbps);) {
+
+            System.out.println("Uploading...");
+            File putFile = new File(putFilepath);
+            PutObjectRequest putRequest = PutObjectRequest.builder().bucket(bucket).key(key)
+                    .contentLength(putFile.length()).build();
+            PutDataSupplier putDataSupplier = new PutGet.PutDataSupplier(putFile);
+            CompletableFuture<PutObjectOutput> putFuture = s3Client.putObject(putRequest, putDataSupplier);
+            try {
+                putFuture.get();
+            } catch (InterruptedException | ExecutionException e) {
+                throw new RuntimeException(e);
+            }
+            System.out.println("Upload complete");
+
+            System.out.println("Downloading...");
+            GetObjectRequest getRequest = GetObjectRequest.builder().bucket(bucket).key(key).build();
+            GetDataConsumer getDataConsumer = new PutGet.GetDataConsumer(getFilepath);
+            CompletableFuture<GetObjectOutput> getFuture = s3Client.getObject(getRequest, getDataConsumer);
+            try {
+                getFuture.get();
+            } catch (InterruptedException | ExecutionException e) {
```

Comment:
extreme nit: ExecutionExceptions usually wrap the relevant exception in the causedBy and don't reveal anything itself. The full stack trace will still show what matters, but usually you can unwrap those once before rewrapping in RuntimeException and throwing that.

Reply:
I'm going to keep it as is, for the sake of keeping the sample as simple as possible. If the stacktrace shows the issue, that's sufficient

Thanks for the info, though